couchbase/src/main/scala/akka/stream/alpakka/couchbase/scaladsl/CouchbaseSession.scala

```diff
@@ -91,192 +93,32 @@ trait CouchbaseSession {
-   *
-   * For replacing other types of documents see `replaceDoc`.
-   *
-   * @return a future that completes when the replace is done
-   */
-  def replace(document: JsonDocument): Future[JsonDocument]
-
-  /**
-   * Replace using the default write settings.
-   *
-   * Separate from `replace` to make the most common case smoother with the type inference
-   *
-   * @return a future that completes when the replace is done
-   */
-  def replaceDoc[T <: Document[_]](document: T): Future[T]
-
-  /**
-   * Replace using the given write settings
-   *
-   * For replacing other types of documents see `replaceDoc`.
-   *
-   * @return a future that completes when the replace is done
-   */
-  def replace(document: JsonDocument, writeSettings: CouchbaseWriteSettings): Future[JsonDocument]
-
-  /**
-   * Replace using the given write settings
-   *
-   * Separate from `replace` to make the most common case smoother with the type inference
-   *
-   * @return a future that completes when the replace is done
-   */
-  def replaceDoc[T <: Document[_]](document: T, writeSettings: CouchbaseWriteSettings): Future[T]
-
-  /**
-   * Remove a document by id using the default write settings.
-   *
-   * @return Future that completes when the document has been removed, if there is no such document
-   *         the future is failed with a `DocumentDoesNotExistException`
-   */
-  def remove(id: String): Future[Done]
-
-  /**
-   * Remove a document by id using the default write settings.
-   *
-   * @return Future that completes when the document has been removed, if there is no such document
-   *         the future is failed with a `DocumentDoesNotExistException`
-   */
-  def remove(id: String, writeSettings: CouchbaseWriteSettings): Future[Done]
-
-  def streamedQuery(query: N1qlQuery): Source[JsonObject, NotUsed]
-  def streamedQuery(query: Statement): Source[JsonObject, NotUsed]
-  def singleResponseQuery(query: Statement): Future[Option[JsonObject]]
-  def singleResponseQuery(query: N1qlQuery): Future[Option[JsonObject]]
-
-  /**
-   * Create or increment a counter
-   *
-   * @param id What counter document id
-   * @param delta Value to increase the counter with if it does exist
-   * @param initial Value to start from if the counter does not exist
-   * @return The value of the counter after applying the delta
-   */
-  def counter(id: String, delta: Long, initial: Long): Future[Long]
-
-  /**
-   * Create or increment a counter
-   *
-   * @param id What counter document id
-   * @param delta Value to increase the counter with if it does exist
-   * @param initial Value to start from if the counter does not exist
-   * @return The value of the counter after applying the delta
-   */
-  def counter(id: String, delta: Long, initial: Long, writeSettings: CouchbaseWriteSettings): Future[Long]
+  private val collectionSessions = new AtomicReference(
+    mutable.WeakHashMap.empty[(String, String), CouchbaseCollectionSession]
+  )
```

Comment:
The general pattern for doing this kind of compare and swap is that the collection inside would be immutable.

I see this is safe since it is probably safe since cloned before adding an entry on each try, but better to use Java concurrent hash map in this case.

WeakHashMap and weak references are problematic for GC in general so I'd recommend not to try to use that as a cache eviction strategy. 

If a pool is really needed it'd be better be done as an akka extension, but I think we should take a step back and consider if this is really worth it. And if it is worth a pool of clients, let's do that a s a separate PR.

```